### ISSUE TITLE
Multiple uuid's which don't exist

### ISSUE THREAD

**Issue by RCheesley** [chaoss/grimoirelab#238]
I have an issue with multiple UUID's being displayed in the Affiliations view, but they don't exist in Hatstall.

Is there some way to refresh or clear out any where this is the case?

I'm not sure how it has occurred but it's clogging up the view and I'm somewhat concerned for the data validity if we have duplication going on.

**Comment by GeorgLink** [comment]
Did you recently merge identities? The data in the dashboard will be updated over time (I think because it has to alter the history of all activities). Until that update is complete, you will continue to see outdated identities (outdated because you merged them with a different identity).

**Comment by RCheesley** [comment]
Ahhh, that could be the case! I had been seeing them for some time but I'll wait and see if it sorts itself out in a bit (it's currently working through updating data sources so maybe this will be resolved once those are done!).

Is it feasible to do some kind of internal lookup for the old UUID once it has been merged, or does hatstall drop the references to the old UI?

**Comment by RCheesley** [comment]
I don't think this is the case now, it's been three days and I've had more data coming in, still some people are not being correctly associated to the company (although others have been successfully) and the UUID's remain.

**Comment by GeorgLink** [comment]
Then I don't know. @valeriocos might be able to help.

**Comment by valeriocos** [comment]
Hi @RCheesley , have you tried to recreate your enriched indexes?

**Comment by RCheesley** [comment]
@valeriocos can you point me at some documentation which explains how to do this please, and I'd gladly try! :)

**Comment by GeorgLink** [comment]
@RCheesley have you tried the command from this section of the GrimoireLab tutorial? 

https://chaoss.github.io/grimoirelab-tutorial/sortinghat/basic.html#regenerating-enriched-indexes

**Comment by valeriocos** [comment]
Sorry for the late reply @RCheesley !!

you can do it from command line:
```
curl -XDELETE <elasticsearc-url>/<name-of-the-enriched-index-to-delete> -k
```

For instance, my elasticsearch is at https://admin:admin@localhost:9200 and the index to delete is `github_chaoss_enrich`, thus the curl command is:
```
curl -XDELETE https://admin:admin@localhost:9200/github_chaoss_enrich -k
```

In case you want to delete all indexes that share the same initial chars you can execute:
```
curl -XDELETE https://admin:admin@localhost:9200/github* -k
```

**Comment by RCheesley** [comment]
Is this the case when using the docker setup?  Also, how do I find the name of the index?  Thanks!

**Comment by valeriocos** [comment]
The indexes are declared in your setup.cfg. Taking https://pastebin.com/rMSpguBy as a reference, the git enriched index is `git_demo_enriched`, the github enriched index is `github_demo_enriched` and so on.

If you want to list all indexes in your elasticsearch, you should execute this command from terminal (or directly from the dev tool of Kibana): `your-elasticsearch-url/_cat/indices?pretty`. A similar result can be obtained also with `your-elasticsearch-url/_alias?pretty`, in this case you will see the indexes and their related aliases if exist.

For example:
```
- curl -XGET https://admin:admin@localhost:9200/_cat/indices?pretty -k
- curl -XGET https://admin:admin@localhost:9200/_alias?pretty -k
```

**Comment by RCheesley** [comment]
OK thanks, appreciate the explanations to a newbie!

**Comment by valeriocos** [comment]
you're welcome @RCheesley !!

**Comment by RCheesley** [comment]
Here's a screencast, this is after completely rebuilding from scratch and pulling in the data sources again, I tried to delete the indexes and rebuild but it persisted.

Some of the screens are showing correct attributions from hatstall, but the git, overview, and the affiliations screen are not, and only show two companies and unknown despite the users being associated with companies.

Quick screencast demonstrating: https://youtu.be/DEkhgWfzRzg

**Comment by valeriocos** [comment]
Thank you @RCheesley for the screencast, this is going to be helpful to understand the problem. We will address your issue ASAP.

**Comment by valeriocos** [comment]
@zhquan and I had a look at your issue. We relied on the configuration files at https://github.com/chaoss/grimoirelab-sirmordred/issues/374. We included some organizations (e.g., `Mautic`, `Acquia`) via sortinghat (see command below), however you can achieve the same results by modifying the orgs_file.
```
- sortinghat -u root --host mariadb -d ruth_sh orgs -a Acquia <---- to add the organization
- sortinghat -u root --host mariadb -d ruth_sh orgs -a Acquia acquia.com <--- to add the domain acquia.com to the organization Acquia
```

 As you can see from the screenshots below git, github and affiliations seem fine.
![image (1)](https://user-images.githubusercontent.com/6515067/69446019-d417d600-0d53-11ea-9878-f10b1396123f.png)
![image (2)](https://user-images.githubusercontent.com/6515067/69446028-d5e19980-0d53-11ea-87bd-d6d216a1311a.png)
![image (3)](https://user-images.githubusercontent.com/6515067/69446032-d8dc8a00-0d53-11ea-9063-a4a0bad1e237.png)

Maybe your issue is related to something wrong in the affiliation file (line 34 here:https://pastebin.com/rMSpguBy)?

**Comment by RCheesley** [comment]
I have an organizations.json and orgs_file in the root of the directory, is that what you mean, or do I need to look within the docker containers themselves?

**Comment by RCheesley** [comment]
(note, most of the organisations I have added via Hatstall manually, rather than with the orgs files)

**Comment by valeriocos** [comment]
>https://github.com/chaoss/grimoirelab/issues/238#issuecomment-557621406

I guess it's ok, i was referring to the orgs_file loaded at line 25 of https://pastebin.com/rMSpguBy.

>https://github.com/chaoss/grimoirelab/issues/238#issuecomment-557621583

When you added the organizations, did you add also the domains? Next week we can try to replicate your issue by inserting orgs from hatstall.

**Comment by RCheesley** [comment]
When you add an organisation in Hatstall, it does not ask you for a domain, only the organisation name:

![screenshot-157 245 245 139_8000-2019 11](https://user-images.githubusercontent.com/2930593/69450382-ee9e7f00-0d54-11ea-8270-525b54e5c63b.png)

**Comment by valeriocos** [comment]
Sorry for the late reply @RCheesley . 

Domains can be added only via SortingHat, please try the following:
- access your server where grimoirelab is running
- execute `docker ps` to get the list of containers
- execute `docker exec -it <mordred-container-ID> bash` to access the container
- within the container, execute `sortinghat -u root -p <password> --host <mariadb-host> -d <database-name> orgs -a <org-name> <org-domain>`

Note that `password`, `mariadb-host` and `database-name` are defined in your setup.cfg.

Example of `docker ps` output:
```
slimbook@slimbook-KATANA:~$ docker ps
CONTAINER ID        IMAGE                                  COMMAND                  CREATED              STATUS              PORTS                              NAMES
9b429b0b4ab5        bitergia/kibiter:secured-v6.1.4-2      "/docker-entrypoint.…"   About a minute ago   Up About a minute   0.0.0.0:5601->5601/tcp             docker_kibiter_1
0fd1add66cda        bitergia/elasticsearch:6.1.0-secured   "/docker-entrypoint.…"   About a minute ago   Up About a minute   8000/tcp, 0.0.0.0:9200->9200/tcp   docker_elasticsearch_1
```

Example of `docker-exec`
```
slimbook@slimbook-KATANA:~$ docker exec -it 9b429b0b4ab5 bash
root@9b429b0b4ab5:/#
```

Example of `sortinghat` command (assuming that the database is `ruth_sh`, there is no password, the host is `mariadb` and the organization `Acquia` already exists).
```
sortinghat -u root --host mariadb -d ruth_sh orgs -a Acquia acquia.com
```

**Comment by RCheesley** [comment]
I think I'm missing something here, what has adding domains got to do with
the UUID's being unknown and profiles not being merged when they are merged
in Hatstall?

Ruth Cheesley

ACQUIA

Mautic Community Manager

p: +44 (0)1442 734282 | m: +44 (0) 7308 164475 | e: ruth.cheesley@acquia.com

/ WE ARE THE OPEN SOURCE DIGITAL EXPERIENCE COMPANY.
<https://www.acquia.com/>

**Comment by valeriocos** [comment]
Let's suppose you create a commit with `ruth.cheesley@acquia.com` and `acquia.com` is a domain registered for the `Acquia` in SortingHat. When the matching operation is executed, the organization `Acquia` is retrieved from `acquia.com` (thanks to the info stored in SortingHat) and automatically set to your profile.

If the connection between `Acquia` and `acquia.com` isn't defined, the profile info cannot be automatically updated by GrimoireLab. Thus, you have to manually affiliate user by user via HatStall.
Please @zhquan correct me if I'm wrong.

There is an issue in Hatstall to allow managing domains, https://github.com/chaoss/grimoirelab-hatstall/issues/68. @sduenas @dlumbrer maybe we could evaluate the effort to implement it, WDYT?

**Comment by RCheesley** [comment]
OK, but the issue is that I have *already* associated the user in Hatstall
with an organisation. Their profiles have already been merged, yet in the
Affiliations screen the old UUID's are still showing from before the merge.

So the user and organisation associations are (apparently) working
correctly, as expected, but we're seeing what seem to be orphaned UUID's
still remaining in the Affiliations screen.

Are you suggesting that every single company I add in Hatstall as an
organisation must have a domain name associated with it and be in the orgs
file? How about companies who maybe don't have a domain (e.g. freelancers)?

Ruth Cheesley

ACQUIA

Mautic Community Manager

p: +44 (0)1442 734282 | m: +44 (0) 7308 164475 | e: ruth.cheesley@acquia.com

/ WE ARE THE OPEN SOURCE DIGITAL EXPERIENCE COMPANY.
<https://www.acquia.com/>

**Comment by valeriocos** [comment]
Sorry for the misunderstanding and late reply, @RCheesley 

> OK, but the issue is that I have *already* associated the user in Hatstall
> with an organisation. Their profiles have already been merged, yet in the
> Affiliations screen the old UUID's are still showing from before the merge.

In theory this should solved in the next enrichement phase, however in your case it doesn't seem to work (https://github.com/chaoss/grimoirelab/issues/238#issuecomment-554651656).

It is difficult to replicate your issue since we don't have access to the data and the test we performed locally (https://github.com/chaoss/grimoirelab/issues/238#issuecomment-557617445) seemed to be successful. 

Can you access SortingHat and report the output of the following command?
```
sortinghat -u root --host mariadb -d ruth_sh show <UUID>
```

**Comment by RCheesley** [comment]
Thanks again!

I think the issue was a delay in the enrichment (several days worth or
delay!) or some kind of problem stalling the process silently.  We had to
rebuild the containers to do some work to move it to SSL and set up
security, and after they were rebuilt the UUID's seemed to start working as
expected and being cleared out when a user was merged.

So we can close this issue now as the problem seemed to be resolved by
rebuilding the containers. None the wiser as to the reason behind the
problem though!

Thanks again for your help and patience!

Ruth

Ruth Cheesley

ACQUIA

Mautic Community Manager

p: +44 (0)1442 734282 | m: +44 (0) 7308 164475 | e: ruth.cheesley@acquia.com

/ WE ARE THE OPEN SOURCE DIGITAL EXPERIENCE COMPANY.
<https://www.acquia.com/>
